syntax = "proto3";

package kvstore;

service KeyValueStore {
    rpc Get(GetRequest) returns (GetResponse) {}
    rpc Put(PutRequest) returns (PutResponse) {}
    rpc appendEntries(AppendRequest) returns (AppendResponse) {}
    rpc requestVote(VoteRequest) returns(VoteResponse){}
}

// You'll likely need to define more specific return codes than these!
enum ReturnCode {
    SUCCESS = 0;
    FAILURE = 1;
}

message GetRequest {
    string key = 1;
}

message GetResponse {
    string value = 1;
    ReturnCode ret = 2;
}

message PutRequest {
    string key = 1;
    string value = 2;
}

message PutResponse {
    ReturnCode ret = 1;
}

enum AppendType {
    HEARTBEAT = 0;
    GET = 1;
    PUT = 2;
}

message LogEntry {
    int32 term;
    PutRequest log;
}

message AppendRequest {
    int32 term = 1;
    int32 leaderID = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    int32 leaderCommit = 6;
    //should include incoming server id
    int32 incServerID = 7;
}

message AppendResponse {
    int32 term = 1;
    bool success = 2;
}

message VoteRequest{
    int32 term = 1;
    int32 candidateID = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

message VoteResponse{
    int32 term = 1;
    bool voteGranted = 2;
}